include/deal.II/matrix_free/matrix_free.templates.h

```diff
@@ -391,21 +411,34 @@ initialize_dof_handlers (const std::vector<const DoFHandler<dim>*> &dof_handler,
 
   dof_info.resize (dof_handlers.n_dof_handlers);
 
-  // go through cells on zeroth level and then successively step down into
+  // Go through cells on zeroth level and then successively step down into
   // children. This gives a z-ordering of the cells, which is beneficial when
   // setting up neighboring relations between cells for thread parallelization
   const unsigned int n_mpi_procs = size_info.n_procs;
   const unsigned int my_pid = size_info.my_pid;
 
+  // If this is a serial Triangulation we simply take all cells,
+  // no matter the subdomain_id.
+  // Otherwise, only consider cells with the appropriate subdomain_id.
   const Triangulation<dim> &tria = dof_handlers.dof_handler[0]->get_triangulation();
+  const parallel::Triangulation<dim> *parallel_triangulation
+    = dynamic_cast<const parallel::Triangulation<dim> *>
+      (&dof_handlers.dof_handler[0]->get_triangulation());
   if (level == numbers::invalid_unsigned_int)
     {
       if (n_mpi_procs == 1)
         cell_level_index.reserve (tria.n_active_cells());
       typename Triangulation<dim>::cell_iterator cell = tria.begin(0),
                                                  end_cell = tria.end(0);
-      for ( ; cell != end_cell; ++cell)
-        internal::resolve_cell (cell, cell_level_index, my_pid);
+      if (parallel_triangulation!=0)
```

Comment:
Since you don't actually use the `parallel_triangulation` variable as a triangulation, and since you only need it here, why not inline the `dynamic_cast` here?